File: training/v1d.py
"""V 本体（`PROCEDURE.md` 2.1 の作るもの #1）。

調波 prior への**加算複素残差**を出す 1D 因果ネット。自由位相にしない
（位相は prior が供給する。励起源を入れながら位相を自由にすると源と競合して
「かすれ」が出るという確定した帰属がある）。

`rddsp_hf.Wavehax2D` は使わない——`nn.GroupNorm(1, ch)` が `[B,C,F,T]` の T 込みで
正規化するので**発話全体統計**になり、`CLAUDE.md` の出荷ゲート（推論経路に発話全体の
統計を置かない）に落ちる。ここは時間軸をまたぐ正規化層を一切置かない。

    入力  cin = N_MEL + 3 * NBIN     （mel / prior real / prior imag / prior log|.|）
      -> Conv1d(cin -> dim, k_in, 左パディングのみ)
      -> ConvNeXtBlock1d(dim, k=k, causal=True) * L
      -> LayerNorm(dim) -> Linear(dim -> 2 * NBIN)
      -> S = complex(P.real + o[0], P.imag + o[1])
"""
from __future__ import annotations

import torch
import torch.nn as nn

import ship_front as SF
from kansei_vocoder import ConvNeXtBlock1d

EPS = 1e-5


def cin_of(nbin: int) -> int:
    return SF.N_MEL + 3 * nbin


def ctx_of(k_in: int, k: int, layers: int) -> int:
    """幹の左文脈（合成フレーム数）。ckpt の `args` に焼いて Rust 側と突き合わせる。"""
    return (k_in - 1) + layers * (k - 1)


class V1D(nn.Module):
    def __init__(self, cin: int | None = None, dim: int = 256, L: int = 6,
                 nbin: int = SF.NFFT_S // 2 + 1, k_in: int = 7, k: int = 3):
        """署名は `PROCEDURE.md` 2.1 の宣言どおり（Rust 側とキー名を一致させる）。

        `cin` は `N_MEL + 3*nbin` から導けるが、**明示させて食い違いを早く落とす**。
        """
        super().__init__()
        want = cin_of(nbin)
        if cin is not None and cin != want:
            raise ValueError(f"cin={cin} は nbin={nbin} と合わない（正 {want}）")
        layers = L
        self.nbin = nbin
        self.dim = dim
        self.layers = layers
        self.k_in = k_in
        self.k = k
        self.cin = cin_of(nbin)
        self.ctx = ctx_of(k_in, k, layers)
        self.inp = nn.Conv1d(self.cin, dim, k_in, padding=0)
        # ⚠ k を明示して渡す。既定 7 のままだと CTX が 42・net 0.26 になり、
        #   27 時間走ってから RTF で落ちる（2.4b-1）。
        self.blocks = nn.ModuleList(
            [ConvNeXtBlock1d(dim, k=k, causal=True) for _ in range(layers)])
        self.norm = nn.LayerNorm(dim)
        self.out = nn.Linear(dim, 2 * nbin)
        nn.init.zeros_(self.out.weight)
        nn.init.zeros_(self.out.bias)

    def features(self, mel_syn: torch.Tensor, P: torch.Tensor) -> torch.Tensor:
        """[N_MEL, T] と複素 prior [NBIN, T] を cin 系列に畳む。"""
        return torch.cat([mel_syn, P.real, P.imag,
                          (P.abs() + EPS).log()], dim=0)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        """**正典の署名**（`PROCEDURE.md` 2.3 (1)）: [B, cin, T] -> [B, 2*NBIN, T]。

        左パディングのみ＝先読み 0。prior への加算は呼び出し側が行う
        （`residual()` / `apply()` が定型を持つ）。
        """
        x = self.inp(nn.functional.pad(x, (self.k_in - 1, 0)))
        for b in self.blocks:
            x = b(x)
        return self.out(self.norm(x.transpose(1, 2))).transpose(1, 2)

    def apply(self, x: torch.Tensor, P: torch.Tensor) -> torch.Tensor:
        """学習経路。x: [B, cin, T], P: [B, NBIN, T] complex -> [B, NBIN, T]。

        **加算複素残差**（自由位相にしない。位相は prior が供給する）。
        """
        o = self(x)
        return torch.complex(P.real + o[:, :self.nbin], P.imag + o[:, self.nbin:])

    def residual(self, mel_syn: torch.Tensor, P: torch.Tensor) -> torch.Tensor:
        """1 発話ぶんの便宜経路。mel_syn: [N_MEL, T], P: [NBIN, T] complex。"""
        return self.apply(self.features(mel_syn, P)[None], P[None])[0]

    def arch(self) -> dict:
        """ckpt の `args` に焼く再構築情報（2.1 の作るもの i）。"""
        return {"arch": "v1d", "dim": self.dim, "L": self.layers,
                "k_in": self.k_in, "k": self.k, "nbin": self.nbin,
                "cin": self.cin, "ctx": self.ctx,
                "nfft_s": SF.NFFT_S, "hop_s": SF.HOP_S,
                "nfft_a": SF.NFFT_A, "hop_a": SF.HOP_A, "n_mel": SF.N_MEL}


class V1DStream:
    """ブロック実行（2.1 の作るもの #2c）。**層ごとに左文脈 `k-1` を保持**する。

    キャッシュ無しで毎ブロック span 全再計算すると net RTF 0.3971（実測）で
    予算 0.207 に到底届かない。それは方式の否定ではなく**未完成実装の否定**なので、
    RTF の判定はこちらで測る（`memory/no-half-baked-implementations`）。

    1 呼び出しで `emit` 個の合成フレームを出し、内部状態だけを進める。
    先読みは 0（右パディングを一切しない）。
    """

    def __init__(self, net: "V1D"):
        self.net = net
        self.reset()

    def reset(self) -> None:
        n = self.net
        self.c_in = torch.zeros(1, n.cin, n.k_in - 1)
        self.c_blk = [torch.zeros(1, n.dim, n.k - 1) for _ in range(n.layers)]

    @torch.no_grad()
    def step(self, mel_syn: torch.Tensor, P: torch.Tensor) -> torch.Tensor:
        """mel_syn: [N_MEL, emit], P: [NBIN, emit] complex -> [NBIN, emit]。"""
        n = self.net
        x = n.features(mel_syn, P).unsqueeze(0)
        x = torch.cat([self.c_in, x], dim=-1)
        self.c_in = x[..., -(n.k_in - 1):] if n.k_in > 1 else self.c_in
        x = n.inp(x)
        for i, b in enumerate(n.blocks):
            h = torch.cat([self.c_blk[i], x], dim=-1)
            self.c_blk[i] = h[..., -(n.k - 1):] if n.k > 1 else self.c_blk[i]
            r = h[..., n.k - 1:]
            y = b.dw(h).transpose(1, 2)
            y = b.norm(y)
            y = b.pw2(b.act(b.pw1(y))).transpose(1, 2)
            x = r + y
        o = n.out(n.norm(x.transpose(1, 2))).transpose(1, 2)[0]
        return torch.complex(P.real + o[:n.nbin], P.imag + o[n.nbin:])

File: crates/lightvc-core/src/v1d.rs
//! V1D — additive complex residual on top of the harmonic prior.
//!
//! Mirrors `training/v1d.py`. Keys match PyTorch exactly (`CLAUDE.md`):
//! `inp.{weight,bias}` / `blocks.{i}.dw.{weight,bias}` / `blocks.{i}.norm.*` /
//! `blocks.{i}.pw1.*` / `blocks.{i}.pw2.*` / `norm.*` / `out.*`.
//!
//! Left padding only — zero lookahead. Phase comes from the prior; the trunk
//! never predicts it freely.

use candle_core::{Module, Result, Tensor, D};
use candle_nn::{Conv1d, Conv1dConfig, LayerNorm, Linear, VarBuilder};

pub const N_MEL: usize = 80;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct V1dCfg {
    pub dim: usize,
    pub layers: usize,
    pub k_in: usize,
    pub k: usize,
    pub nbin: usize,
}

impl V1dCfg {
    pub const DEFAULT: V1dCfg = V1dCfg { dim: 256, layers: 6, k_in: 7, k: 3, nbin: 257 };

    #[inline]
    pub fn cin(&self) -> usize {
        N_MEL + 3 * self.nbin
    }

    /// Left context of the trunk, in synthesis frames.
    #[inline]
    pub fn ctx(&self) -> usize {
        (self.k_in - 1) + self.layers * (self.k - 1)
    }
}

fn conv1d_plain(i: usize, o: usize, k: usize, vb: VarBuilder) -> Result<Conv1d> {
    let w = vb.get((o, i, k), "weight")?;
    let b = vb.get((o,), "bias")?;
    Ok(Conv1d::new(w, Some(b), Conv1dConfig::default()))
}

fn linear(i: usize, o: usize, vb: VarBuilder) -> Result<Linear> {
    let w = vb.get((o, i), "weight")?;
    let b = vb.get((o,), "bias")?;
    Ok(Linear::new(w, Some(b)))
}

fn layer_norm(dim: usize, vb: VarBuilder) -> Result<LayerNorm> {
    let w = vb.get((dim,), "weight")?;
    let b = vb.get((dim,), "bias")?;
    Ok(LayerNorm::new(w, b, 1e-5))
}

struct Block {
    dw: Conv1d,
    norm: LayerNorm,
    pw1: Linear,
    pw2: Linear,
    k: usize,
}

impl Block {
    fn new(cfg: &V1dCfg, vb: VarBuilder) -> Result<Self> {
        Ok(Self {
            dw: conv1d_plain(cfg.dim, cfg.dim, cfg.k, vb.pp("dw"))?,
            norm: layer_norm(cfg.dim, vb.pp("norm"))?,
            pw1: linear(cfg.dim, cfg.dim * 3, vb.pp("pw1"))?,
            pw2: linear(cfg.dim * 3, cfg.dim, vb.pp("pw2"))?,
            k: cfg.k,
        })
    }

    /// `h` already carries the k-1 frames of left context; returns [B,dim,T].
    fn tail(&self, h: &Tensor, r: &Tensor) -> Result<Tensor> {
        let y = self.dw.forward(h)?;
        let y = y.transpose(1, 2)?.contiguous()?;
        let y = self.norm.forward(&y)?;
        let y = self.pw2.forward(&self.pw1.forward(&y)?.gelu_erf()?)?;
        r + &y.transpose(1, 2)?
    }

    fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let h = x.pad_with_zeros(D::Minus1, self.k - 1, 0)?;
        self.tail(&h, x)
    }
}

pub struct V1d {
    cfg: V1dCfg,
    inp: Conv1d,
    blocks: Vec<Block>,
    norm: LayerNorm,
    out: Linear,
}

impl V1d {
    pub fn load(cfg: V1dCfg, vb: VarBuilder) -> Result<Self> {
        let inp = conv1d_plain(cfg.cin(), cfg.dim, cfg.k_in, vb.pp("inp"))?;
        let mut blocks = Vec::with_capacity(cfg.layers);
        for i in 0..cfg.layers {
            blocks.push(Block::new(&cfg, vb.pp(format!("blocks.{i}")))?);
        }
        let norm = layer_norm(cfg.dim, vb.pp("norm"))?;
        let out = linear(cfg.dim, 2 * cfg.nbin, vb.pp("out"))?;
        Ok(Self { cfg, inp, blocks, norm, out })
    }

    pub fn cfg(&self) -> V1dCfg {
        self.cfg
    }

    /// Trunk. `x`: [B, cin, T] -> [B, 2*nbin, T]. Canonical signature.
    pub fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let h = x.pad_with_zeros(D::Minus1, self.cfg.k_in - 1, 0)?;
        let mut h = self.inp.forward(&h)?;
        for b in &self.blocks {
            h = b.forward(&h)?;
        }
        let h = h.transpose(1, 2)?.contiguous()?;
        let h = self.out.forward(&self.norm.forward(&h)?)?;
        h.transpose(1, 2)
    }

    /// Concatenate the trunk input the same way `V1D.features` does.
    pub fn features(mel: &Tensor, pre: &Tensor, pim: &Tensor, plog: &Tensor) -> Result<Tensor> {
        Tensor::cat(&[mel, pre, pim, plog], 1)
    }

    /// Additive complex residual. Returns (real, imag), each [B, nbin, T].
    pub fn apply(&self, x: &Tensor, pre: &Tensor, pim: &Tensor) -> Result<(Tensor, Tensor)> {
        let o = self.forward(x)?;
        let nb = self.cfg.nbin;
        Ok((
            (pre + o.narrow(1, 0, nb)?)?,
            (pim + o.narrow(1, nb, nb)?)?,
        ))
    }
}

/// Per-layer left-context caches (block execution). Mirrors `V1DStream`.
pub struct V1dStream {
    c_in: Tensor,
    c_blk: Vec<Tensor>,
}

impl V1dStream {
    pub fn new(net: &V1d, dev: &candle_core::Device) -> Result<Self> {
        let c = net.cfg;
        Ok(Self {
            c_in: Tensor::zeros((1, c.cin(), c.k_in - 1), candle_core::DType::F32, dev)?,
            c_blk: (0..c.layers)
                .map(|_| Tensor::zeros((1, c.dim, c.k - 1), candle_core::DType::F32, dev))
                .collect::<Result<Vec<_>>>()?,
        })
    }

    /// Emit `T` frames; advances state. `x`: [1, cin, T].
    pub fn step(&mut self, net: &V1d, x: &Tensor, pre: &Tensor, pim: &Tensor)
        -> Result<(Tensor, Tensor)>
    {
        let c = net.cfg;
        let h = Tensor::cat(&[&self.c_in, x], D::Minus1)?;
        let n = h.dim(D::Minus1)?;
        self.c_in = h.narrow(D::Minus1, n - (c.k_in - 1), c.k_in - 1)?.contiguous()?;
        let mut y = net.inp.forward(&h)?;
        for (i, b) in net.blocks.iter().enumerate() {
            let hh = Tensor::cat(&[&self.c_blk[i], &y], D::Minus1)?;
            let m = hh.dim(D::Minus1)?;
            self.c_blk[i] = hh.narrow(D::Minus1, m - (c.k - 1), c.k - 1)?.contiguous()?;
            let r = hh.narrow(D::Minus1, c.k - 1, m - (c.k - 1))?.contiguous()?;
            y = b.tail(&hh, &r)?;
        }
        let y = y.transpose(1, 2)?.contiguous()?;
        let o = net.out.forward(&net.norm.forward(&y)?)?.transpose(1, 2)?;
        let nb = c.nbin;
        Ok(((pre + o.narrow(1, 0, nb)?)?, (pim + o.narrow(1, nb, nb)?)?))
    }
}

// ---------------------------------------------------------------------------
// C.3 parity tests (torch ≡ candle ≤ 1e-4, streaming ≡ offline ≥ 80 dB)
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use candle_core::{DType, Device};

    fn dir() -> std::path::PathBuf {
        std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join("testdata")
    }

    fn read_f32(name: &str) -> Vec<f32> {
        let b = std::fs::read(dir().join(name)).expect("testdata");
        b.chunks_exact(4)
            .map(|c| f32::from_le_bytes([c[0], c[1], c[2], c[3]]))
            .collect()
    }

    fn load(dev: &Device) -> (V1d, V1dCfg) {
        let cfg = V1dCfg::DEFAULT;
        let vb = unsafe {
            VarBuilder::from_mmaped_safetensors(
                &[dir().join("v1d_test.safetensors")], DType::F32, dev).unwrap()
        };
        (V1d::load(cfg, vb).unwrap(), cfg)
    }

    #[test]
    fn torch_parity() {
        let dev = Device::Cpu;
        let (net, cfg) = load(&dev);
        let t = 32usize;
        let x = Tensor::from_vec(read_f32("tv_x.bin"), (1, cfg.cin(), t), &dev).unwrap();
        let pre = Tensor::from_vec(read_f32("tv_pre.bin"), (1, cfg.nbin, t), &dev).unwrap();
        let pim = Tensor::from_vec(read_f32("tv_pim.bin"), (1, cfg.nbin, t), &dev).unwrap();
        let (re, im) = net.apply(&x, &pre, &pim).unwrap();
        for (got, want) in [(re, "tv_re.bin"), (im, "tv_im.bin")] {
            let g = got.flatten_all().unwrap().to_vec1::<f32>().unwrap();
            let w = read_f32(want);
            let e = g.iter().zip(&w).map(|(a, b)| (a - b).abs()).fold(0.0f32, f32::max);
            assert!(e <= 1e-4, "torch != candle: max abs err {e} ({want})");
        }
    }

    #[test]
    fn streaming_equals_offline() {
        let dev = Device::Cpu;
        let (net, cfg) = load(&dev);
        let t = 32usize;
        let x = Tensor::from_vec(read_f32("tv_x.bin"), (1, cfg.cin(), t), &dev).unwrap();
        let pre = Tensor::from_vec(read_f32("tv_pre.bin"), (1, cfg.nbin, t), &dev).unwrap();
        let pim = Tensor::from_vec(read_f32("tv_pim.bin"), (1, cfg.nbin, t), &dev).unwrap();
        let (fre, _) = net.apply(&x, &pre, &pim).unwrap();

        let mut st = V1dStream::new(&net, &dev).unwrap();
        let mut parts = Vec::new();
        let blk = 2usize;
        for i in (0..t).step_by(blk) {
            let xs = x.narrow(D::Minus1, i, blk).unwrap();
            let ps = pre.narrow(D::Minus1, i, blk).unwrap();
            let qs = pim.narrow(D::Minus1, i, blk).unwrap();
            parts.push(st.step(&net, &xs, &ps, &qs).unwrap().0);
        }
        let sre = Tensor::cat(&parts, D::Minus1).unwrap();

        // ⚠ 頭 (NFFT_S - HOP_S)/HOP_S = 3 合成フレームを除外して測る（C.3）。
        let skip = 3usize;
        let a = sre.narrow(D::Minus1, skip, t - skip).unwrap()
            .flatten_all().unwrap().to_vec1::<f32>().unwrap();
        let b = fre.narrow(D::Minus1, skip, t - skip).unwrap()
            .flatten_all().unwrap().to_vec1::<f32>().unwrap();
        let (mut e, mut s) = (0f64, 0f64);
        for (p, q) in a.iter().zip(&b) {
            e += ((p - q) as f64).powi(2);
            s += (*q as f64).powi(2);
        }
        let snr = 10.0 * (s / e.max(1e-30)).log10();
        assert!(snr >= 80.0, "streaming != offline: SNR {snr:.2} dB");
    }

    #[test]
    fn zero_lookahead() {
        // 時刻 c 以降の入力を書き換えても、c 以前の出力は 1 bit も動かない。
        let dev = Device::Cpu;
        let (net, cfg) = load(&dev);
        let t = 32usize;
        let x = Tensor::from_vec(read_f32("tv_x.bin"), (1, cfg.cin(), t), &dev).unwrap();
        let pre = Tensor::from_vec(read_f32("tv_pre.bin"), (1, cfg.nbin, t), &dev).unwrap();
        let pim = Tensor::from_vec(read_f32("tv_pim.bin"), (1, cfg.nbin, t), &dev).unwrap();
        let base = net.apply(&x, &pre, &pim).unwrap().0;

        let c = 20usize;
        let head = x.narrow(D::Minus1, 0, c).unwrap();
        let tail = Tensor::ones((1, cfg.cin(), t - c), DType::F32, &dev).unwrap();
        let x2 = Tensor::cat(&[&head, &tail], D::Minus1).unwrap();
        let alt = net.apply(&x2, &pre, &pim).unwrap().0;

        let a = base.narrow(D::Minus1, 0, c).unwrap()
            .flatten_all().unwrap().to_vec1::<f32>().unwrap();
        let b = alt.narrow(D::Minus1, 0, c).unwrap()
            .flatten_all().unwrap().to_vec1::<f32>().unwrap();
        let e = a.iter().zip(&b).map(|(p, q)| (p - q).abs()).fold(0.0f32, f32::max);
        assert!(e == 0.0, "先読みがある: 頭 {c} フレームが {e} 動いた");
    }
}
